### PR TITLE
Add project_share_group resource to share projects with groups

### PR DESCRIPTION
**Pull request by enieuw**
This resources allows you to share a project with a group. Not really fond of the name but it follows the description as given here: https://docs.gitlab.com/ee/api/projects.html#share-project-with-group

### CODE REVIEW

**Comment by gordonbondon** [comment]
Does not `shared_with_groups` parameter on `gitlab_project` resource already solve this?

**Comment by enieuw** [comment]
It can do the same yes but it requires the project to be in the state of terraform. This decouples the linking resources similar to resources like aws security groups with inline security rules and separate rules.

On 7 Aug 2019, at 20:23, Artem Iarmoliuk <notifications@github.com<mailto:notifications@github.com>> wrote:


Does not shared_with_groups parameter on gitlab_project resource already solve this?

File: gitlab/resource_gitlab_project_share_group.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		if groupId == v.GroupID {
+			resourceGitlabProjectShareGroupSetToState(d, v, &projectId)
+		}
+	}
+
+	return nil
+}
+
+func projectIdAndGroupIdFromId(id string) (string, int, error) {
+	projectId, groupIdString, err := parseTwoPartID(id)
```
Comment (roidelapluie) [inline gitlab/resource_gitlab_project_share_group.go]: return directly if err != nil

File: gitlab/resource_gitlab_project_share_group.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	d.SetId(buildTwoPartID(&projectId, &groupIdString))
+	return resourceGitlabProjectShareGroupRead(d, meta)
+}
+
+func resourceGitlabProjectShareGroupRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	id := d.Id()
+	log.Printf("[DEBUG] read gitlab project projectMember %s", id)
+
+	projectId, groupId, e := projectIdAndGroupIdFromId(id)
```
Comment (roidelapluie) [inline gitlab/resource_gitlab_project_share_group.go]: use `err` for errors

File: gitlab/resource_gitlab_project_share_group.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		return err
+	}
+	return resourceGitlabProjectShareGroupRead(d, meta)
+}
+
+func resourceGitlabProjectShareGroupDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	id := d.Id()
+	projectId, groupId, e := projectIdAndGroupIdFromId(id)
```
Comment (roidelapluie) [inline gitlab/resource_gitlab_project_share_group.go]: use `err` for errors

File: gitlab/resource_gitlab_project_share_group.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabProjectShareGroup() *schema.Resource {
+	acceptedAccessLevels := make([]string, 0, len(accessLevelID))
```
Comment (roidelapluie) [inline gitlab/resource_gitlab_project_share_group.go]: len -1 if we exclude owner?

File: gitlab/resource_gitlab_project_share_group.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	for k := range accessLevelID {
+		if k != "owner" {
```
Comment (roidelapluie) [inline gitlab/resource_gitlab_project_share_group.go]: is it needed?

File: gitlab/resource_gitlab_project_share_group.go (hunk 4 above)
Comment (enieuw) [reply gitlab/resource_gitlab_project_share_group.go]: Changed it, I now made this a hardcoded list as there were more invalid options in the accessLevelID list. The only options in the hardcoded list now are options which will actually get accepted by the Gitlab API

File: gitlab/resource_gitlab_project_share_group.go (hunk 5 above)
Comment (enieuw) [reply gitlab/resource_gitlab_project_share_group.go]: Same as above, hardcoded the list

**Comment by enieuw** [comment]
Processed your comments, I copied quite a bit from the project membership resource which is where the non err variables came from.  Should be resolved now.

**Pull request merged**